lib/Imagine/Factory/ClassFactory.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+namespace Imagine\Factory;
+
+use Imagine\Exception\InvalidArgumentException;
+use Imagine\Exception\RuntimeException;
+use Imagine\File\Loader;
+use Imagine\Image\Box;
+use Imagine\Image\ImageInterface;
+use Imagine\Image\Metadata\DefaultMetadataReader;
+use Imagine\Image\Metadata\ExifMetadataReader;
+use Imagine\Image\Metadata\MetadataBag;
+use Imagine\Image\Palette\Color\ColorInterface;
+use Imagine\Image\Palette\PaletteInterface;
+
+/**
+ * The default implementation of Imagine\Factory\ClassFactoryInterface.
+ */
+class ClassFactory implements ClassFactoryInterface
+{
+    /**
+     * @var array|null
+     */
+    private static $gdInfo;
+
+    /**
+     * {@inheritdoc}
+     *
+     * @see \Imagine\Factory\ClassFactoryInterface::createMetadataReader()
+     */
+    public function createMetadataReader()
+    {
+        return $this->finalize(ExifMetadataReader::isSupported() ? new ExifMetadataReader() : new DefaultMetadataReader());
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @see \Imagine\Factory\ClassFactoryInterface::createBox()
+     */
+    public function createBox($width, $height)
+    {
+        return $this->finalize(new Box($width, $height));
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @see \Imagine\Factory\ClassFactoryInterface::createFileLoader()
+     */
+    public function createFileLoader($path)
+    {
+        return $this->finalize(new Loader($path));
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @see \Imagine\Factory\ClassFactoryInterface::createDrawer()
+     */
+    public function createDrawer($handle, $resource)
+    {
+        switch ($handle) {
+            case static::HANDLE_GD:
```

Comment:
you should use `self::` instead of relying on late static binding. Overriding the constants is not a supported extension point anyway (as the code is likely to use the ones from the interface)

Reply:
👍 done in f00fe304343f0657ce51a80275c7bb227afa57a7